expression/builtin_string.go

```diff
@@ -2974,20 +2974,31 @@ func (b *builtinFormatSig) Clone() builtinFunc {
 	return newSig
 }
 
+// formatMaxDecimals limits the maximum number of decimal digits for result of
+// function `format`, this value is same as `FORMAT_MAX_DECIMALS` in MySQL source code.
+const formatMaxDecimals int64 = 30
+
 // evalString evals FORMAT(X,D).
 // See https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_format
 func (b *builtinFormatSig) evalString(row chunk.Row) (string, bool, error) {
-	x, isNull, err := b.args[0].EvalString(b.ctx, row)
+	x, isNull, err := b.args[0].EvalReal(b.ctx, row)
```

Comment:
The argument types are all `ETString`: https://github.com/pingcap/tidb/pull/8797/files#diff-314e997a9df9b116e8f0aad4149df468R2920. We should not change `EvalString()` to `EvalReal()`.

Reply:
This this the output of MySQL:
```
mysql> select format("1a","2a");
+-------------------+
| format("1a","2a") |
+-------------------+
| 1.00              |
+-------------------+
1 row in set, 2 warnings (0.00 sec)

mysql> show warnings;
+---------+------+-----------------------------------------+
| Level   | Code | Message                                 |
+---------+------+-----------------------------------------+
| Warning | 1292 | Truncated incorrect INTEGER value: '2a' |
| Warning | 1292 | Truncated incorrect DOUBLE value: '1a'  |
+---------+------+-----------------------------------------+
2 rows in set (0.00 sec)
```
Seems MySQL believes the type of parameters to be `Double` and `Integer`, but it is not reflected in https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_format.

Reply:
In MySQL source code, for function `format(X, D)`, `X` is evaluated as decimal or real type, while `D` is evaluated as integer type.
```cpp
  dec= (int) args[1]->val_int();
  if (args[0]->result_type() == DECIMAL_RESULT ||
      args[0]->result_type() == INT_RESULT)
  {
    res= args[0]->val_decimal(&dec_val);
  }
  else
  {
    double nr= args[0]->val_real();
  }
```

Reply:
seems we should modify https://github.com/pingcap/tidb/pull/8797/files#diff-314e997a9df9b116e8f0aad4149df468R2920 to change the argument type.